src/evaluators/source-4-3.js

```diff
@@ -0,0 +1,808 @@
+// The values of primitive functions are "primitive"
+// objects, see line 281 how such functions are applied
+
+function setup_environment() {
+    const primitive_function_names =
+        map(f => head(f), primitive_functions);
+    const primitive_function_values =
+        map(f => make_primitive_function(head(tail(f))),
+            primitive_functions);
+    const primitive_constant_names =
+        map(f => head(f), primitive_constants);
+    const primitive_constant_values =
+        map(f => head(tail(f)),
+            primitive_constants);
+    return extend_environment(
+        append(primitive_function_names,
+            primitive_constant_names),
+        append(primitive_function_values,
+            primitive_constant_values),
+        the_empty_environment);
+}
+
+const the_global_environment = setup_environment();
+
+// below are amb related
+function is_amb(stmt) {
+    return is_tagged_list(stmt, "application") &&
+        is_name(operator(stmt)) &&
+        name_of_name(operator(stmt)) === "amb";
+}
+
+function amb_choices(stmt) {
+    return operands(stmt);
+}
+
+function ambeval(exp, env, succeed, fail) {
+    return analyze(exp)(env, succeed, fail);
+}
+
+function analyze_self_evaluating(stmt) {
+    return (env, succeed, fail) => succeed(stmt, fail);
+}
+
+function analyze_name(stmt) {
+    return (env, succeed, fail) =>
+        succeed(lookup_name_value(name_of_name(stmt), env),
+            fail);
+}
+
+function analyze_function_definition(stmt) {
+    const vars =
+        function_definition_parameters(stmt);
+    const bfun =
+        analyze(function_definition_body(stmt));
+    const locals = local_names(function_definition_body(stmt));
+    return (env, succeed, fail) =>
+        succeed(make_compound_function(vars, bfun, locals, env),
+            fail);
+}
+
+function analyze_conditional_expression(stmt) {
+    const pfun = analyze(cond_expr_pred(stmt));
+    const cfun = analyze(cond_expr_cons(stmt));
+    const afun = analyze(cond_expr_alt(stmt));
+    return (env, succeed, fail) =>
+        pfun(env,
+            // success continuation for evaluating the predicate
+            // to obtain pred_value
+            (pred_value, fail2) =>
+                is_true(pred_value) ?
+                    cfun(env, succeed, fail2) :
+                    afun(env, succeed, fail2),
+            fail);
+}
+
+function analyze_sequence(stmts) {
+    function sequentially(a, b) {
+        return (env, succeed, fail) =>
+            a(env,
+                (a_value, fail2) => b(env, succeed, fail2),
+                fail);
+    }
+
+    function loop(first_fun, rest_funs) {
+        return is_null(rest_funs) ?
+            first_fun :
+            loop(sequentially(first_fun,
+                head(rest_funs)),
+                tail(rest_funs));
+    }
+    const funs = map(analyze, stmts);
+    return is_null(funs) ?
+        env => undefined :
+        loop(head(funs), tail(funs));
+}
+
+// ??
```

Comment:
What does // ?? mean?

Reply:
This was for our own book keeping, removed in a15a077